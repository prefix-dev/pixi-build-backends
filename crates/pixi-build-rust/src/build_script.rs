use minijinja::Environment;
use serde::Serialize;

#[derive(Serialize)]
pub struct BuildScriptContext {
    /// The location of the source
    pub source_dir: String,

    /// Any additional args to pass to `cargo`
    pub extra_args: Vec<String>,

    /// True if `openssl` is part of the build environment
    pub has_openssl: bool,

    /// True if `sccache` is available.
    pub has_sccache: bool,

    /// The platform that is running the build.
    pub is_bash: bool,
}

impl BuildScriptContext {
    pub fn render(&self) -> Vec<String> {
        let env = Environment::new();
        let template = env
            .template_from_str(include_str!("build_script.j2"))
            .unwrap();
        let rendered = template.render(self).unwrap().to_string();
        rendered
            .lines()
            .map(|s| s.to_string())
            .filter(|s| !s.is_empty())
            .collect()
    }
}

#[cfg(test)]
mod test {
    use rstest::*;

    #[rstest]
    fn test_build_script(#[values(true, false)] is_bash: bool) {
        let context = super::BuildScriptContext {
            source_dir: String::from("my-prefix-dir"),
            extra_args: vec![],
            has_openssl: false,
            has_sccache: false,
            is_bash,
        };
        let script = context.render();

        let mut settings = insta::Settings::clone_current();
        settings.set_snapshot_suffix(if is_bash { "bash" } else { "cmdexe" });
        settings.bind(|| {
            insta::assert_snapshot!(script.join("\n"));
        });
    }

    #[rstest]
    fn test_sccache(#[values(true, false)] is_bash: bool) {
        let context = super::BuildScriptContext {
            source_dir: String::from("my-prefix-dir"),
            extra_args: vec![],
            has_openssl: false,
            has_sccache: true,
            is_bash,
        };
        let script = context.render();

        let mut settings = insta::Settings::clone_current();
        settings.set_snapshot_suffix(if is_bash { "bash" } else { "cmdexe" });
        settings.bind(|| {
            insta::assert_snapshot!(script.join("\n"));
        });
    }

    #[rstest]
    fn test_openssl(#[values(true, false)] is_bash: bool) {
        let context = super::BuildScriptContext {
            source_dir: String::from("my-prefix-dir"),
            extra_args: vec![],
            has_openssl: true,
            has_sccache: false,
            is_bash,
        };
        let script = context.render();

        let mut settings = insta::Settings::clone_current();
        settings.set_snapshot_suffix(if is_bash { "bash" } else { "cmdexe" });
        settings.bind(|| {
            insta::assert_snapshot!(script.join("\n"));
        });
    }
}
