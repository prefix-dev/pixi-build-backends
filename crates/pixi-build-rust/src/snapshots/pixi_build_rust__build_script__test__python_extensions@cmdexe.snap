---
source: crates/pixi-build-rust/src/build_script.rs
expression: "script.join(\"\\n\")"
---
cargo build --release --manifest-path my-prefix-dir/Cargo.toml 
if errorlevel 1 exit 1
if exist "my-prefix-dir\target\release\*.exe" (
    for %%f in ("my-prefix-dir\target\release\*.exe") do (
        copy "%%f" "%PREFIX%\bin\"
    )
)
if exist "my-prefix-dir\target\release\mymodule.dll" (
    mkdir "%PREFIX%\Lib\site-packages" 2>nul
    copy "my-prefix-dir\target\release\mymodule.dll" "%PREFIX%\Lib\site-packages\mymodule.pyd"
) else if exist "my-prefix-dir\target\release\libmymodule.dll" (
    mkdir "%PREFIX%\Lib\site-packages" 2>nul
    copy "my-prefix-dir\target\release\libmymodule.dll" "%PREFIX%\Lib\site-packages\mymodule.pyd"
)
if exist "my-prefix-dir\target\release\anothermodule.dll" (
    mkdir "%PREFIX%\Lib\site-packages" 2>nul
    copy "my-prefix-dir\target\release\anothermodule.dll" "%PREFIX%\Lib\site-packages\anothermodule.pyd"
) else if exist "my-prefix-dir\target\release\libanothermodule.dll" (
    mkdir "%PREFIX%\Lib\site-packages" 2>nul
    copy "my-prefix-dir\target\release\libanothermodule.dll" "%PREFIX%\Lib\site-packages\anothermodule.pyd"
)
